File: backend/app/api/v1/schemas/dependency.py
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict

from app.db.models.dependency import DependencyType, DependencySource, DependencyDirection, HttpMethod
from app.api.v1.schemas.system import SystemResponse, SubSystemResponse


# ---------------------------------------------------------------------------
# Nested base schemas for dependency responses
# ---------------------------------------------------------------------------


class SystemBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    description: Optional[str] = None


class SubSystemBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    description: Optional[str] = None
    system_id: int


# ---------------------------------------------------------------------------
# SystemDependency schemas
# ---------------------------------------------------------------------------


class SystemDependencyCreate(BaseModel):
    to_system_id: int
    dependency_type: DependencyType
    direction: DependencyDirection = DependencyDirection.ONE_WAY
    source: DependencySource = DependencySource.MANUAL
    label: Optional[str] = None


class SystemDependencyUpdate(BaseModel):
    dependency_type: Optional[DependencyType] = None
    source: Optional[DependencySource] = None
    direction: Optional[DependencyDirection] = None
    label: Optional[str] = None


class SystemDependencyResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    from_system_id: int
    to_system_id: int
    dependency_type: DependencyType
    direction: DependencyDirection
    source: DependencySource
    label: Optional[str] = None
    tenant_id: int
    to_system: SystemResponse
    from_system: Optional[SystemBase] = None
    is_incoming: bool = False


# ---------------------------------------------------------------------------
# ComponentEndpoint schemas
# ---------------------------------------------------------------------------


class ComponentEndpointCreate(BaseModel):
    http_method: Optional[HttpMethod] = None
    path: str
    description: Optional[str] = None


class ComponentEndpointUpdate(BaseModel):
    http_method: Optional[HttpMethod] = None
    path: Optional[str] = None
    description: Optional[str] = None


class ComponentEndpointResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    dependency_id: int
    http_method: Optional[HttpMethod] = None
    path: str
    description: Optional[str] = None
    tenant_id: int


# ---------------------------------------------------------------------------
# ComponentDependency schemas
# ---------------------------------------------------------------------------


class ComponentDependencyCreate(BaseModel):
    to_subsystem_id: int
    dependency_type: DependencyType
    direction: DependencyDirection = DependencyDirection.ONE_WAY
    protocol: Optional[str] = None
    port: Optional[int] = None
    source: DependencySource = DependencySource.MANUAL
    label: Optional[str] = None


class ComponentDependencyUpdate(BaseModel):
    dependency_type: Optional[DependencyType] = None
    protocol: Optional[str] = None
    port: Optional[int] = None
    source: Optional[DependencySource] = None
    direction: Optional[DependencyDirection] = None
    label: Optional[str] = None


class ComponentDependencyResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    from_subsystem_id: int
    to_subsystem_id: int
    dependency_type: DependencyType
    direction: DependencyDirection
    protocol: Optional[str] = None
    port: Optional[int] = None
    source: DependencySource
    label: Optional[str] = None
    tenant_id: int
    to_subsystem: SubSystemResponse
    from_subsystem: Optional[SubSystemBase] = None
    is_incoming: bool = False
    endpoints: list[ComponentEndpointResponse] = []


# ---------------------------------------------------------------------------
# Verify response schemas
# ---------------------------------------------------------------------------


class DependencyVerifyItem(BaseModel):
    to_system_id: int
    to_system_name: str
    dependency_type: DependencyType
    status: Literal["satisfied", "mocked", "missing"]


class SystemVerifyResult(BaseModel):
    system_id: int
    system_name: str
    dependencies: list[DependencyVerifyItem]


class ComponentVerifyItem(BaseModel):
    from_subsystem_id: int
    from_subsystem_name: str
    to_subsystem_id: int
    to_subsystem_name: str
    dependency_type: DependencyType
    status: Literal["satisfied", "mocked", "missing"]


class VerifyResponse(BaseModel):
    environment_id: int
    total_dependencies: int
    satisfied_count: int
    mocked_count: int
    missing_count: int
    systems: list[SystemVerifyResult]
    component_total: int = 0
    component_satisfied: int = 0
    component_mocked: int = 0
    component_missing: int = 0
    component_dependencies: list[ComponentVerifyItem] = []

File: frontend/src/types/dependency.ts
export type DependencyType = 'api_call' | 'database' | 'message_queue' | 'event' | 'file' | 'other';
export type HttpMethod = 'get' | 'post' | 'put' | 'patch' | 'delete' | 'head' | 'options';

export interface ComponentEndpointResponse {
  id: number;
  dependency_id: number;
  http_method: HttpMethod | null;
  path: string;
  description: string | null;
  tenant_id: number;
}

export interface ComponentEndpointCreate {
  http_method?: HttpMethod | null;
  path: string;
  description?: string | null;
}

export interface ComponentEndpointUpdate {
  http_method?: HttpMethod | null;
  path?: string;
  description?: string | null;
}
export type DependencySource = 'manual' | 'terraform' | 'docker_compose';
export type DependencyDirection = 'one_way' | 'two_way';

export interface SystemDependencyResponse {
  id: number;
  from_system_id: number;
  to_system_id: number;
  dependency_type: DependencyType;
  direction: DependencyDirection;
  source: DependencySource;
  tenant_id: number;
  to_system: {
    id: number;
    name: string;
    description: string | null;
  };
  from_system?: {
    id: number;
    name: string;
    description?: string;
  };
  is_incoming: boolean;
}

export interface ComponentDependencyResponse {
  id: number;
  from_subsystem_id: number;
  to_subsystem_id: number;
  dependency_type: DependencyType;
  direction: DependencyDirection;
  label: string | null;
  protocol: string | null;
  port: number | null;
  source: DependencySource;
  tenant_id: number;
  to_subsystem: {
    id: number;
    name: string;
    description: string | null;
    system_id: number;
  };
  from_subsystem?: {
    id: number;
    name: string;
    description?: string;
    system_id: number;
  };
  is_incoming: boolean;
  endpoints: ComponentEndpointResponse[];
}

export interface SystemDependencyCreate {
  to_system_id: number;
  dependency_type: DependencyType;
  direction?: DependencyDirection;
  source?: DependencySource;
}

export interface ComponentDependencyCreate {
  to_subsystem_id: number;
  dependency_type: DependencyType;
  direction?: DependencyDirection;
  protocol?: string;
  port?: number;
  source?: DependencySource;
}

export interface DependencyVerifyItem {
  to_system_id: number;
  to_system_name: string;
  dependency_type: DependencyType;
  status: 'satisfied' | 'mocked' | 'missing';
}

export interface SystemVerifyResult {
  system_id: number;
  system_name: string;
  dependencies: DependencyVerifyItem[];
}

export interface ComponentVerifyItem {
  from_subsystem_id: number;
  from_subsystem_name: string;
  to_subsystem_id: number;
  to_subsystem_name: string;
  dependency_type: DependencyType;
  status: 'satisfied' | 'mocked' | 'missing';
}

export interface VerifyResponse {
  environment_id: number;
  total_dependencies: number;
  satisfied_count: number;
  mocked_count: number;
  missing_count: number;
  systems: SystemVerifyResult[];
  component_total: number;
  component_satisfied: number;
  component_mocked: number;
  component_missing: number;
  component_dependencies: ComponentVerifyItem[];
}
